nupic/network/synapse.py

```diff
@@ -0,0 +1,38 @@
+﻿# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+
+class Synapse(object):

```

Comment:
Expanding on my contrived example, consider the following python code:

```python
import sys



ITERATIONS = 10000000



class MutableObject(object):
  def __init__(self, foo=None):
    self.foo = foo



class SlottedMutableObject(MutableObject):
  __slots__ = ("foo",)



def create():
  for index in xrange(ITERATIONS):
    obj = MutableObject("bar")



def mutate(objects = {}):
  for index in xrange(ITERATIONS):
    obj = objects.setdefault(index, MutableObject("bar"))



def mutateSlotted(objects = {}):
  for index in xrange(ITERATIONS):
    obj = objects.setdefault(index, SlottedMutableObject("bar"))



if __name__ == "__main__":
  selection = sys.argv[1]

  if selection == "create":
    create()
  elif selection == "mutate":
    mutate()
  elif selection == "mutate-slotted":
    mutateSlotted()
```

`create()` creates 10,000,000 `MutableObject` instances, `mutate()` instantiates and caches them, `mutateSlotted()` does the same with a subclass that defines `__slots__`.  When I run each of those with `cProfile`, here's what I get:

```
Austins-MacBook-Pro:Desktop amarshall$ python -m cProfile mutant.py create
         10000005 function calls in 3.851 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    3.851    3.851 mutant.py:1(<module>)
 10000000    1.145    0.000    1.145    0.000 mutant.py:10(__init__)
        1    0.000    0.000    0.000    0.000 mutant.py:15(SlottedMutableObject)
        1    2.706    2.706    3.851    3.851 mutant.py:20(create)
        1    0.000    0.000    0.000    0.000 mutant.py:9(MutableObject)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


Austins-MacBook-Pro:Desktop amarshall$ python -m cProfile mutant.py mutate
         20000005 function calls in 15.278 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   15.278   15.278 mutant.py:1(<module>)
 10000000    5.746    0.000    5.746    0.000 mutant.py:10(__init__)
        1    0.000    0.000    0.000    0.000 mutant.py:15(SlottedMutableObject)
        1    8.372    8.372   15.278   15.278 mutant.py:26(mutate)
        1    0.000    0.000    0.000    0.000 mutant.py:9(MutableObject)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
 10000000    1.160    0.000    1.160    0.000 {method 'setdefault' of 'dict' objects}


Austins-MacBook-Pro:Desktop amarshall$ python -m cProfile mutant.py mutate-slotted
         20000005 function calls in 10.649 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   10.649   10.649 mutant.py:1(<module>)
 10000000    0.986    0.000    0.986    0.000 mutant.py:10(__init__)
        1    0.000    0.000    0.000    0.000 mutant.py:15(SlottedMutableObject)
        1    8.568    8.568   10.649   10.649 mutant.py:32(mutateSlotted)
        1    0.000    0.000    0.000    0.000 mutant.py:9(MutableObject)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
 10000000    1.095    0.000    1.095    0.000 {method 'setdefault' of 'dict' objects}
```

Notice that `create()` is ~5x faster than `mutate()`.  That's just for the _creation_ of the objects.  It is faster to do a lookup in a `dict` than it is to create an object, however, but there's memory overhead in retaining the `dict` in which the objects are cached and as the results above show, there's performance overhead in caching the objects in a dict.  When I add `print sys.getsizeof(objects)` to the end of `mutate()` and `mutateSlotted()` I get 402653464 bytes.

Are we willing to take the memory hit?  Probably not, but it would be cool if it were an option so that the user can make that choice according to their own usage pattern.

Something else to consider, any memory overhead that would need to be translated to a persisted checkpoint could potentially slow things down in a production setting, such as is the case in `htmengine` where models are checkpointed and loaded from disk regularly.